### PR TITLE
Changes property options to support `converter`

### LINKED ISSUE #264: Discussion: supported serializers out of the box
I recently wrote [this blog post](https://43081j.com/2018/10/lit-element-types) about the fact that we don't support `type: Object` and `type: Array` like Polymer did. I also made [this repository](https://github.com/43081j/lit-element-serializers).

This is a thing people assume quite often, [even in the recent docs PR](https://github.com/Polymer/lit-element/blob/e205f51b67bda647001248afa43545b28c8135a4/docs/projects/try/expressions/custom-element.js#L6).

So I think we need to make a decision before 1.0, one of the following two choices:

* Ship serializers for `Object` and `Array` (and possibly require that consumers import them rather than supporting them by default)
* Somehow make it clearer that `type` is _not_ a type constructor but rather a type serializer (at this point we can't really rename `type` but maybe we could document it _very well_)

I'm happy to PR this if we go with the former.

Right now, people almost always assume because they see examples use `type: Number`, `type: String`, etc, they should then use `type: Date`, `type: Object`, `type: Array` and so on. This won't work as expected, but people rarely bump into this realisation as they pass their data via properties rather than attributes.

We do want to keep the core as light as we can, so if we do introduce this, it probably belongs in an extra module we must import ourselves.

cc @justinfagnani

### PR DESCRIPTION
**Pull request by sorvell**
Fixes #264

Changes `type` to be only a hint to the `converter` option which has the previous `type` functionality, an object with `toAttribute` and `fromAttribute` or just a function which is `fromAttribute`. In addition to the `value` these functions now also get the property's `type`.

Also provides a default converter that supports `Boolean`, `String`, `Number`, `Object`, and `Array` out of the box.

In addition, numbers and strings now become `null` if their reflected attribute is removed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,11 +35,12 @@ for additional information on how to create templates for lit-element.
...
     If a string, the string value is observed (e.g `attribute: 'foo-bar'`).
-    * `type`: Indicates how to serialize and deserialize the attribute to/from a property.
+    * `converter`: Indicates how to serialize and deserialize the attribute to/from a property.
     The value can be a function used for both serialization and deserialization, or it can
     be an object with individual functions via the optional keys, `fromAttribute` and `toAttribute`.
-    `type` defaults to the `String` constructor, and so does the `toAttribute` and `fromAttribute`
-    keys.
+    * `type`: Indicates the type of the property. This is used only as a hint for the
+    `converter` to determine how to serialize and deserialize the attribute
+    to/from a property.
```
Comment (kevinpschaaf) [inline README.md]: Add: 
> Default converters are provided for `String`, `Number`, ...

File: src/lib/updating-element.ts [hunk 2]
```diff
@@ -90,9 +99,33 @@ type AttributeMap = Map<string, PropertyKey>;
...
-const toBooleanAttribute = (value: string) => value ? '' : null;
+export const defaultConverter: ComplexAttributeConverter = {
+
+  toAttribute(value: any, type?: any) {
+    switch (type) {
+    case Boolean:
+      return value ? '' : null;
+    case Object:
+    case Array:
+      return JSON.stringify(value);
```
Comment (kevinpschaaf) [inline src/lib/updating-element.ts]: Based on experience with this being a footgun for users, consider anything better?

File: src/lib/updating-element.ts [hunk 3]
```diff
@@ -260,21 +294,16 @@ export abstract class UpdatingElement extends HTMLElement {
...
+   * `converter` or `converter.fromAttribute` property option.
    */
   private static _propertyValueFromAttribute(value: string,
                                              options?: PropertyDeclaration) {
     const type = options && options.type;
-    if (type === undefined) {
-      return value;
-    }
-    // Note: special case `Boolean` so users can use it as a `type`.
+    const converter = options && options.converter || defaultConverter;
```
Comment (kevinpschaaf) [inline src/lib/updating-element.ts]: Shouldn't need a null check since there's a `defaultPropertyDeclaration`?

File: README.md (hunk 1 above)
Comment (sorvell) [reply README.md]: Added to `converter` section.

File: README.md [hunk 4]
```diff
@@ -35,11 +35,16 @@ for additional information on how to create templates for lit-element.
...
     be an object with individual functions via the optional keys, `fromAttribute` and `toAttribute`.
-    `type` defaults to the `String` constructor, and so does the `toAttribute` and `fromAttribute`
-    keys.
+    A default `converter` is used if none is provided; it supports
+    `Boolean`, `String`, `Number`, `Object`, and `Array`.
+    * `type`: Indicates the type of the property. This is used only as a hint for the
+    `converter` to determine how to serialize and deserialize the attribute
+    to/from a property. Note, when a property changes and the converter is used
+    to update the attribute, the property is never updated again as a result of
+    the attribute changing, and visa versa.
```
Comment (vicb) [inline README.md]:
```suggestion
    the attribute changing, and vice versa.
```

File: src/lib/updating-element.ts [hunk 5]
```diff
@@ -15,27 +15,28 @@
 /**
  * Converts property values to and from attribute values.
  */
-export interface AttributeSerializer<T = any> {
+export interface ComplexAttributeConverter<Type = any, TypeHint = any> {
 
   /**
    * Deserializing function called to convert an attribute value to a property
```
Comment (vicb) [inline src/lib/updating-element.ts]: IMO `Deserializing` and `Deserializing` is a confusing terminology.
Is it still needed after this change ?

File: README.md (hunk 4 above)
Comment (vicb) [inline README.md]: `docs/_guide/properties.md` should also be updated, see #370

File: src/lib/updating-element.ts (hunk 2 above)
Comment (sorvell) [reply src/lib/updating-element.ts]: Added this #375 to ensure this is well documented.

File: src/lib/updating-element.ts [hunk 6]
```diff
@@ -490,12 +523,13 @@ export abstract class UpdatingElement extends HTMLElement {
...
-    // track old value when changing.
+    // Track old value when changing.
     if (!this._changedProperties.has(name)) {
       this._changedProperties.set(name, oldValue);
     }
-    // add to reflecting properties set
-    if (options.reflect === true) {
+    // Add to reflecting properties set if `reflect` is true and the property
+    // is not reflecting from the attribute
+    if (options.reflect === true &&  !(this._updateState & STATE_IS_REFLECTING)) {
```
[the remark this answers is not in the document]
Comment (sorvell) [reply src/lib/updating-element.ts]: Done.

File: src/lib/updating-element.ts [hunk 7]
```diff
@@ -416,41 +446,44 @@ export abstract class UpdatingElement extends HTMLElement {
       name: PropertyKey, value: unknown,
       options: PropertyDeclaration = defaultPropertyDeclaration) {
     const ctor = (this.constructor as typeof UpdatingElement);
-    const attrValue = ctor._propertyValueToAttribute(value, options);
-    if (attrValue !== undefined) {
```
[the remark this answers is not in the document]
Comment (sorvell) [reply src/lib/updating-element.ts]: Done and added test.

File: src/lib/updating-element.ts (hunk 5 above)
Comment (sorvell) [reply src/lib/updating-element.ts]: Done.

File: README.md (hunk 4 above)
Comment (sorvell) [reply README.md]: Done.

Comment (sorvell) [reply README.md]: Added this #375 to ensure this is well documented.

**Comment by carsonpowers** [comment]
> In addition, numbers and strings now become null if their reflected attribute is removed.

Shouldn't numbers and strings become `undefined` if they are removed? Changing them to `null` means that defaults set by the component will now be bypassed:

```
function func(foo='bar'){
  console.log(foo);
}

func(null); // null
func(undefined); // 'bar'
```

**Pull request merged**